fix mirroring for multiple cephblockpools

Nikhil-Ladha opened the pull request:
fix rbd mirroring and peer secrets addition, when more than one cephblockpool is configured in ramen tests

Nikhil-Ladha commented:
I am not sure how much aligned this change would be wrt to the drenv tool, but me as a user found this inconvenience while using the tool for my testing and thought of sending a fix for it :)

Nikhil-Ladha commented:
/cc @nirs

Nikhil-Ladha commented:
Test results:
```
Disable rbd-mirror debug logs in cluster 'dr1'
Disable rbd-mirror debug logs in cluster 'dr2'
Getting mirroring info site name for cluster 'dr1'
'cephblockpools.ceph.rook.io/replicapool' output='jsonpath={.status.mirroringInfo.site_name}' found in 0.14 seconds
Getting rbd mirror boostrap peer secret name for cluster 'dr1'
Getting secret pool-peer-token-replicapool token for cluster 'dr1'
'cephblockpools.ceph.rook.io/replicapool-2' output='jsonpath={.status.mirroringInfo.site_name}' found in 0.17 seconds
Getting rbd mirror boostrap peer secret name for cluster 'dr1'
Getting secret pool-peer-token-replicapool-2 token for cluster 'dr1'
Getting mirroring info site name for cluster 'dr2'
'cephblockpools.ceph.rook.io/replicapool' output='jsonpath={.status.mirroringInfo.site_name}' found in 0.16 seconds
Getting rbd mirror boostrap peer secret name for cluster 'dr2'
Getting secret pool-peer-token-replicapool token for cluster 'dr2'
'cephblockpools.ceph.rook.io/replicapool-2' output='jsonpath={.status.mirroringInfo.site_name}' found in 0.16 seconds
Getting rbd mirror boostrap peer secret name for cluster 'dr2'
Getting secret pool-peer-token-replicapool-2 token for cluster 'dr2'
Setting up mirroring from 'dr2' to 'dr1'
Applying rbd mirror secret in cluster 'dr1'
secret/a1846069-412a-48d6-9427-bb7d03c3668b configured
Configure peers for cluster 'dr1'
cephblockpool.ceph.rook.io/replicapool patched
secret/a1846069-412a-48d6-9427-bb7d03c3668b configured
Configure peers for cluster 'dr1'
cephblockpool.ceph.rook.io/replicapool-2 patched
Creating VolumeReplicationClass
volumereplicationclass.replication.storage.openshift.io/vrc-sample unchanged
volumegroupreplicationclass.replication.storage.openshift.io/vgrc-sample unchanged
Apply rbd mirror to cluster 'dr1'
cephrbdmirror.ceph.rook.io/my-rbd-mirror unchanged
Setting up mirroring from 'dr1' to 'dr2'
Applying rbd mirror secret in cluster 'dr2'
secret/af74e308-8008-4c4f-a18f-53c7d228239c configured
Configure peers for cluster 'dr2'
cephblockpool.ceph.rook.io/replicapool patched
secret/af74e308-8008-4c4f-a18f-53c7d228239c configured
Configure peers for cluster 'dr2'
cephblockpool.ceph.rook.io/replicapool-2 patched
Creating VolumeReplicationClass
volumereplicationclass.replication.storage.openshift.io/vrc-sample unchanged
volumegroupreplicationclass.replication.storage.openshift.io/vgrc-sample unchanged
Apply rbd mirror to cluster 'dr2'
cephrbdmirror.ceph.rook.io/my-rbd-mirror unchanged
Waiting until rbd mirror is ready in cluster 'dr1'
'cephrbdmirror/my-rbd-mirror' output='jsonpath={.status.phase}' found in 0.14 seconds
cephrbdmirror.ceph.rook.io/my-rbd-mirror condition met
Cluster 'dr1' rbd mirror status:
  observedGeneration: 1
  phase: Ready

Waiting until rbd mirror is ready in cluster 'dr2'
'cephrbdmirror/my-rbd-mirror' output='jsonpath={.status.phase}' found in 0.12 seconds
cephrbdmirror.ceph.rook.io/my-rbd-mirror condition met
Cluster 'dr2' rbd mirror status:
  observedGeneration: 1
  phase: Ready

Waiting for mirroring health in cluster 'dr1' (1/3)
Cluster 'dr1' mirroring status': {'daemon_health': 'WARNING', 'health': 'WARNING', 'image_health': 'OK', 'states': {}}
Cluster 'dr1' mirroring status': {'daemon_health': 'WARNING', 'health': 'WARNING', 'image_health': 'OK', 'states': {}}
Cluster 'dr1' mirroring status': {'daemon_health': 'WARNING', 'health': 'WARNING', 'image_health': 'OK', 'states': {}}
Cluster 'dr1' mirroring status': {'daemon_health': 'OK', 'health': 'OK', 'image_health': 'OK', 'states': {}}
Cluster 'dr1' mirroring healthy in 68.54 seconds
Cluster 'dr1' ceph block pool status:
  info:
    failureDomain: host
    rbdMirrorBootstrapPeerSecretName: pool-peer-token-replicapool
    type: Replicated
  mirroringInfo:
    lastChanged: '2025-02-13T13:56:24Z'
    lastChecked: '2025-02-13T13:57:24Z'
    mode: image
    peers:
    - client_name: client.rbd-mirror-peer
      direction: rx-tx
      mirror_uuid: 92828599-7e64-47a1-9a81-81803aba320c
      site_name: a1846069-412a-48d6-9427-bb7d03c3668b
      uuid: 987c1ff7-18ee-4dc2-b8c4-3eddf4d150f3
    site_name: af74e308-8008-4c4f-a18f-53c7d228239c
  mirroringStatus:
    lastChecked: '2025-02-13T13:57:24Z'
    summary:
      daemon_health: OK
      health: OK
      image_health: OK
      states: {}
  observedGeneration: 3
  phase: Ready
  poolID: 10
  snapshotScheduleStatus:
    lastChanged: '2025-02-13T13:56:24Z'
    lastChecked: '2025-02-13T13:57:24Z'
    snapshotSchedules:
    - image: '-'
      items:
      - interval: 2m
        start_time: 14:00:00-05:00
      namespace: '-'
      pool: replicapool

Cluster 'dr1' ceph block pool status:
  info:
    failureDomain: host
    rbdMirrorBootstrapPeerSecretName: pool-peer-token-replicapool-2
    type: Replicated
  mirroringInfo:
    lastChanged: '2025-02-13T13:55:34Z'
    lastChecked: '2025-02-13T13:56:35Z'
    mode: image
    peers:
    - client_name: client.rbd-mirror-peer
      direction: rx-tx
      site_name: a1846069-412a-48d6-9427-bb7d03c3668b
      uuid: 7fd8acae-a6f4-45c1-96f2-e354ab731554
    site_name: af74e308-8008-4c4f-a18f-53c7d228239c
  mirroringStatus:
    lastChecked: '2025-02-13T13:56:35Z'
    summary:
      daemon_health: WARNING
      health: WARNING
      image_health: OK
      states: {}
  observedGeneration: 3
  phase: Ready
  poolID: 11
  snapshotScheduleStatus:
    lastChanged: '2025-02-13T13:55:34Z'
    lastChecked: '2025-02-13T13:56:35Z'
    snapshotSchedules:
    - image: '-'
      items:
      - interval: 2m
        start_time: 14:00:00-05:00
      namespace: '-'
      pool: replicapool-2

Waiting for mirroring health in cluster 'dr2' (1/3)
Cluster 'dr2' mirroring status': {'daemon_health': 'OK', 'health': 'OK', 'image_health': 'OK', 'states': {}}
Cluster 'dr2' mirroring healthy in 0.13 seconds
Cluster 'dr2' ceph block pool status:
  info:
    failureDomain: host
    rbdMirrorBootstrapPeerSecretName: pool-peer-token-replicapool
    type: Replicated
  mirroringInfo:
    lastChanged: '2025-02-13T13:56:04Z'
    lastChecked: '2025-02-13T13:57:05Z'
    mode: image
    peers:
    - client_name: client.rbd-mirror-peer
      direction: rx-tx
      mirror_uuid: b073fffc-d0a8-4a5e-9d9e-1b32e32c92f0
      site_name: af74e308-8008-4c4f-a18f-53c7d228239c
      uuid: 9ac1b4cf-3544-4590-abdb-75d29a154ed5
    site_name: a1846069-412a-48d6-9427-bb7d03c3668b
  mirroringStatus:
    lastChecked: '2025-02-13T13:57:05Z'
    summary:
      daemon_health: OK
      health: OK
      image_health: OK
      states: {}
  observedGeneration: 3
  phase: Ready
  poolID: 12
  snapshotScheduleStatus:
    lastChanged: '2025-02-13T13:56:04Z'
    lastChecked: '2025-02-13T13:57:05Z'
    snapshotSchedules:
    - image: '-'
      items:
      - interval: 2m
        start_time: 14:00:00-05:00
      namespace: '-'
      pool: replicapool

Cluster 'dr2' ceph block pool status:
  info:
    failureDomain: host
    rbdMirrorBootstrapPeerSecretName: pool-peer-token-replicapool-2
    type: Replicated
  mirroringInfo:
    lastChanged: '2025-02-13T13:56:15Z'
    lastChecked: '2025-02-13T13:57:16Z'
    mode: image
    peers:
    - client_name: client.rbd-mirror-peer
      direction: rx-tx
      mirror_uuid: 803bd7fa-3616-46d4-9b46-cf2693c278c2
      site_name: af74e308-8008-4c4f-a18f-53c7d228239c
      uuid: 557d1654-fe3e-468b-8c05-44f73d78339a
    site_name: a1846069-412a-48d6-9427-bb7d03c3668b
  mirroringStatus:
    lastChecked: '2025-02-13T13:57:16Z'
    summary:
      daemon_health: OK
      health: OK
      image_health: OK
      states: {}
  observedGeneration: 3
  phase: Ready
  poolID: 13
  snapshotScheduleStatus:
    lastChanged: '2025-02-13T13:56:15Z'
    lastChecked: '2025-02-13T13:57:16Z'
    snapshotSchedules:
    - image: '-'
      items:
      - interval: 2m
        start_time: 14:00:00-05:00
      namespace: '-'
      pool: replicapool-2

Mirroring was setup successfully
```

nirs commented:
> fix rbd mirroring and peer secrets addition, when more than one cephblockpool is configured in ramen tests

Can you open an issue explaining the problem you experience? Our e2e tests works fine with current code since we don't use the second replica pool.

Nikhil-Ladha commented on test/addons/rbd-mirror/start:
```diff
@@ -18,7 +18,7 @@ from drenv import ceph
 from drenv import kubectl
 
-POOL_NAME = "replicapool"
+POOL_NAMES = ["replicapool", "replicapool-2"]
```
Yeah

Nikhil-Ladha commented on test/addons/rbd-mirror/start:
```diff
@@ -32,33 +32,35 @@ def fetch_secret_info(cluster):
     info = {}
 
     print(f"Getting mirroring info site name for cluster '{cluster}'")
```
Ack, will update this.

Nikhil-Ladha commented on test/addons/rbd-mirror/start:
```diff
@@ -86,33 +88,37 @@ def disable_rbd_mirror_debug_logs(cluster):
 
     template = drenv.template("start-data/vgrc-sample.yaml")
-    yaml = template.substitute(cluster=cluster, pool=POOL_NAME)
+    yaml = template.substitute(cluster=cluster, pool="replicapool")
```
I was not sure if we need multiple vrc/vgrc, hence I omitted their creation.
They might just be idle and not be required at all, hence I beleive it might not be worth creating them at the first place and if anybody needs it, they can create it manually.
It is easier to create than a blockpool with mirroring enabled ;)

> But in any case, use POOL_NAMES[0] so if we change the name of the pools this code will remain correct.

Sure, will update that.

nirs replied:
If we call this in parallel the vrc/vgrc will be created automatically. I don't see reason not to create them if we create a pool and set up mirroring, this is the easy part. It will enable interesting tests that we don't do now.

Nikhil-Ladha replied:
I was wondering if we go ahead with creating multiple vrc/vgrc then how are planning to name them? If we just `strip` the number from `replicapool-2` even then are we sure that is how we are going to name always?

Nikhil-Ladha commented on test/addons/rbd-mirror/start:
```diff
@@ -182,24 +189,25 @@ def watch_pool_mirroring_status(cluster, timeout=180):
+                        print(
+                            f"Cluster '{cluster}' mirroring healthy in {elapsed:.2f} seconds"
+                        )
+                        return
```
Indeed, from my testing while the loop reaches for the second pool it hardly takes a couple of secs to complete through as the mirroring already got setup when it was iterating through the first loop. So, probably we can skip the trouble of parallelizing here.

nirs replied:
parallelizing is easy if we do this in the top level, see test/addons/volsync/test.

nirs commented:
> I won't describe it as an issue exactly, but more of an incovenience. I have been using the tool for some time now for VGR testing and haven't pulled the latest code for a long time. A couple of days back I did so, and recreated my env. That's when I saw that we are now creating multiple pools so I thought of using the other pool for some other test and I noticed that though mirroring is enabled on the pool but the setup is not done. And, I had to tweak the code to get it done. That's why I thought of sending this PR, if someone else like me comes across this scenario it could be helpful for them :)

Ok, than no issue needed.

nirs commented:
@Nikhil-Ladha can you update the commit message and pr message to reflect the actual change? (we have multiple pool but configure mirroring only one one, why you want to configure mirroring on the second, etc).

Nikhil-Ladha commented:
> Thanks for the PR!
> 
> One problem in this approach is that we have to change lot of code to loop over pools, and it may not efficient since we need to wait for one pool and then wait for the other.
> 
> In test/addons/volsync/test we use another approach - we run the same test twice using `concurrent.futures.ThreadPoolExecutor`. This keep the code simple and run everything in parallel without modifying the code. I think this is the right approach for new code since it is much easier to work with the code this way, and it is likely more efficient.
> 
> Regardless, we need to include the pool names in logs so we can understand whats going on and debug this.

Updated the code to parallelize the setup of mirroring 2 pools. Also, added pool names in the log wherever applicable.
Please check the code now and let me know if something could be improved.

parikshithb commented on test/addons/rbd-mirror/start:
```diff
@@ -227,6 +238,26 @@ def restart_rbd_mirror_daemon(cluster):
+    cluster1_info = fetch_secret_info(cluster1, pool)
+    cluster2_info = fetch_secret_info(cluster2, pool)
+
+    print(f"Setting up mirroring from '{cluster2}' to '{cluster1}'")
```
log "Setting up mirroring for '{pool}' from...."

parikshithb commented on test/addons/rbd-mirror/start:
```diff
@@ -227,6 +238,26 @@ def restart_rbd_mirror_daemon(cluster):
+    print(f"Setting up mirroring from '{cluster2}' to '{cluster1}'")
+    configure_rbd_mirroring(cluster1, pool, cluster2_info)
+
+    print(f"Setting up mirroring from '{cluster1}' to '{cluster2}'")
```
Same here

parikshithb commented on test/addons/rook-pool/start:
```diff
@@ -43,38 +43,40 @@ def deploy(cluster):
+            "--namespace=rook-ceph",
+            context=cluster,
+        )
+        info = {"ceph pool status": json.loads(out)}
```
log ceph pool status with pool name

parikshithb reviewed (commented):
minor comments wrt to logging

nirs commented:
Seems that this change does a lot of changes already in #1823. Lets wait until that PR is finished.
